### PR TITLE
Improve data loading performance

### PR DESCRIPTION
**Pull request by ziw-liu**
### Problem

Data loading during training was slow. Most of the time is spent on I/O and augmentation.

### Performance tweaks

- Disabled MONAI's metadata tracking (see tutorial [here](https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb))
- Allow decompressing the dataset ahead-of-time and caching the result on local temporary storage
  - This will create a new zarr store in local  `/tmp` (or Windows equivalent) if a dataset of the same name does not exist
  - Initialized chunks in an existing cache store will be skipped (need to manually delete cache if the same store name now has different data)
- Stop normalizing source channels (e.g. phase), where absolute intensity carries physical meaning (#221)

### Behavior changes and fixes

- Fixed training/validation dataset split, now it's done on the FOV level (used to happen on the sliding window level)
- Removed the duplicate `architecture` parameter in config files
- Added a new `data.caching` parameter (default to `false`)
- Added a new `model.log_num_samples` parameter (default to `8`)
  - Sample image are now logged for the first sample of the firstm`model.log_num_samples` batches

### Result

After enabling caching, 64 data-loading workers can saturate an A100 GPU with $B \times C \times D \times W \times H = 32 \times 2 \times 5 \times 512 \times 512$ batches. Training on 300 FOVs (80/20 split) now takes 5 min/epoch. 

Epoch/hour:
image

I have not investigated the impact of system RAM on file system caching performance. During the above test a very large amount of RAM ($1536 \times 0.5 = 768$ GB, decompressed dataset is 500 GB) was available for ZFS caching.

### CODE REVIEW

File: micro_dl/light/engine.py [hunk 1]
```diff
@@ -44,9 +45,11 @@ def __init__(
         self.loss_function = loss_function if loss_function else F.mse_loss
         self.lr = lr
         self.schedule = schedule
+        self.log_num_samples = log_num_samples
         self.training_step_outputs = []
         self.validation_step_outputs = []
-        self.example_input_array = torch.rand(1, 1, 5, 256, 256)  # required to log the graph
+        # required to log the graph
+        self.example_input_array = torch.rand(1, 1, 5, 256, 256)
```
Comment (mattersoflight) [inline micro_dl/light/engine.py:52]: @ziw-liu this line leads to the following output on terminal. I added this so that lightning can log the graph.
  | Name  | Type    | Params | In sizes            | Out sizes          
------------------------------------------------------------------------------
0 | model | Unet25d | 2.0 M  | [1, 1, 5, 256, 256] | [1, 1, 1, 256, 256]
------------------------------------------------------------------------------

This should be `config[data][yx_patchsize]`.
I can't figure out how to access the config parameters inside LightningModule at runtime. Can you look into this?

Comment (ziw-liu) [reply micro_dl/light/engine.py]: See https://github.com/mehta-lab/microDL/pull/232/commits/ca56416453620a14d5dd94ce40f1a11860761bb5

File: micro_dl/light/data.py [hunk 2]
```diff
@@ -109,50 +110,90 @@ def __init__(
...
         self.split_ratio = split_ratio
         self.yx_patch_size = yx_patch_size
         self.augment = augment
+        self.caching = caching
+
+    def _cache(self, lazy_plate: Plate) -> Plate:
+        self.tmp_zarr = os.path.join(
+            tempfile.gettempdir(), os.path.basename(self.data_path)
+        )
+        logging.info(f"Caching dataset at {self.tmp_zarr}.")
```
Comment (mattersoflight) [inline micro_dl/light/data.py:135]: I didn't catch this info on the terminal. It will be useful to point the logger to the same directory as used by `lightning_logs`, i.e., `default_root_dir/lightning_logs/yyyymmdd-hhmmss`.

File: micro_dl/light/data.py [hunk 3]
```diff
@@ -1,73 +1,81 @@
...
     RandAffined,
     RandGaussianSmoothd,
     RandWeightedCropd,
 )
-from torch.utils.data import DataLoader, Dataset, Subset
+from numpy.typing import NDArray
+from torch.utils.data import DataLoader, Dataset
+
+
+class NormalizeTargetd(MapTransform):
```
Comment (mattersoflight) [inline micro_dl/light/data.py:25]: The code is working well. Please add more verobse signatures, docstrings, and comments to entire module.

**Review by mattersoflight — COMMENTED** [review]
The new data loading pipeline improves the training speed by 3x. I *think* this is without caching. I could not see on the console where the temporary file is created. 

I was even able to try out EarlyStopping call back with useful results (observations on confluence).

Suggestions before the PR is merged:
* Write log from data module along with lightning_logs.
* Document data.py more completely. 
* Debug why tensorboard is not visible from the  script to generate plots and movies.

**Review by mattersoflight — APPROVED** [review]

File: micro_dl/light/data.py (hunk 3 above)
Comment (ziw-liu) [reply micro_dl/light/data.py:28]: See 55aabae6e3cdd91f2aa402af85031636d522ad78

**Comment by ziw-liu** [comment]
> Debug why tensorboard is not visible from the script to generate plots and movies.

The `load_event_file` function does not exist in the namespace. Importing valid attributes work.

We can implement this feature in another PR.
